src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(value,null, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(value, tail, null);
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size != index) {
+            checkIndex(index);
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            relinkWithNewNode(value, index);
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        unlink(node);
+        return node.item;
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(T value) {
+        int index = getIndex(value);
+        if (index != -1) {
+            remove(index);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        Node(E element, Node<E> prev, Node<E> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("The index is out of bounds for the list size:"
+                    + size);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> node = null;
+        if (size != 0) {
+            node = runLoopFromHead(index);
+            node = node == null ? runLoopFromTail(index) : node;
```

Comment:
you misunderstood the comment
```suggestion
if (index <= size / 2) {
                node = runLoopFromHead(index)
            } else {
                node = runLoopFromTail(index)
```